saleor/core/management/commands/fix_undiscounted_total_prices.py

```diff
@@ -0,0 +1,157 @@
+
+    WHERE (
+        NOT ("order_orderline"."undiscounted_total_price_gross_amount" =
+                ("order_orderline"."undiscounted_unit_price_gross_amount" *
+                "order_orderline"."quantity")) OR
+        NOT ("order_orderline"."undiscounted_total_price_net_amount" =
+                ("order_orderline"."undiscounted_unit_price_net_amount" *
+                "order_orderline"."quantity")))
+    RETURNING "order_orderline"."order_id";
+"""  # noqa: E501
+
+
+def queryset_in_batches(queryset):
+    """Slice a queryset into batches.
+
+    Input queryset should be sorted be pk.
+    """
+    start_pk = 0
+
+    while True:
+        qs = queryset.filter(pk__gt=start_pk)[:ORDERS_UPDATE_BATCH_SIZE]
+        pks = list(qs.values_list("pk", flat=True))
+
+        if not pks:
+            break
+
+        yield pks
+
+        start_pk = pks[-1]
+
+
+def update_order_undiscounted_price():
+    """Fix orders with discount applied on order lines.
+
+    When the order has a voucher discount applied on lines, it is not visible
+    on the order's undiscounted total price. This method is fixing such orders.
+    """
+
+    # Take orders that has applied lines voucher discounts, but the discount is
+    # not visible in undiscounted price.
+    orders_to_update = Order.objects.filter(
+        Exists(
+            OrderLine.objects.filter(
+                order_id=OuterRef("id"), voucher_code__isnull=False
+            )
+        ),
+        total_gross_amount=F("undiscounted_total_gross_amount"),
+    ).order_by("id")
+
+    updated_orders_pks = []
+    for batch_pks in queryset_in_batches(orders_to_update):
+        orders = Order.objects.filter(pk__in=batch_pks)
+        lines = OrderLine.objects.filter(order_id__in=orders.values("id")).values(
+            "order_id",
+            "undiscounted_total_price_gross_amount",
+            "total_price_gross_amount",
+            "undiscounted_total_price_net_amount",
+            "total_price_net_amount",
+        )
+        lines_discount_data = defaultdict(lambda: (0, 0))
+        for data in lines:
+            discount_amount_gross = (
+                data["undiscounted_total_price_gross_amount"]
+                - data["total_price_gross_amount"]
+            )
+            discount_amount_net = (
+                data["undiscounted_total_price_net_amount"]
+                - data["total_price_net_amount"]
+            )
+            current_discount_gross, current_discount_net = lines_discount_data[
+                data["order_id"]
+            ]
+            lines_discount_data[data["order_id"]] = (
+                current_discount_gross + discount_amount_gross,
+                current_discount_net + discount_amount_net,
+            )
+
+        for order in orders:
+            discount_amount_gross, discount_amount_net = lines_discount_data.get(
+                order.id
+            )
+            if discount_amount_gross > 0 or discount_amount_net > 0:
```

Comment:
Is there a case where `discount_amount_gross` could be negative but `discount_amount_net` positive value?

Reply:
It doesn't make sense, a discount is a discount, you would have to probably do something weird like set the percentage to negative but I don't think that's even possible.